test/aio5gc/context/sessionManagement.go

```diff
@@ -0,0 +1,210 @@
+	return c.dataNetwork
+}
+
+func (c *SmContext) SetUserLocation(location models.NrLocation) {
+	c.userLocation = location
+}
+
+func (c *SmContext) SetPti(pti uint8) {
+	c.pti = pti
+}
+
+func (c *SmContext) GetPti() uint8 {
+	return c.pti
+}
+
+func (c *SmContext) SetPduSessionType(sType uint8) {
+	c.sessionType = sType
+}
+
+func (c *SmContext) GetPduSessionType() uint8 {
+	return c.sessionType
+}
+
+func (c *SmContext) GetSessionRule() *models.SessionRule {
+	return c.sessionRule
+}
+
+func (c *SmContext) SetSessionRule(sessionRule *models.SessionRule) {
+	c.sessionRule = sessionRule
+}
+
+func (c *SmContext) GetDefQosQFI() uint8 {
+	return c.defQosQFI
+}
+
+func (c *SmContext) SetDefQosQFI(defQosQFI uint8) {
+	c.defQosQFI = defQosQFI
+}
+
+func (smContext *SmContext) PDUAddressToNAS() ([12]byte, uint8) {
+	var addr [12]byte
+	var addrLen uint8
+	copy(addr[:], smContext.pduAddress)
+	switch smContext.sessionType {
+	case nasMessage.PDUSessionTypeIPv4:
+		addrLen = 4 + 1
+	case nasMessage.PDUSessionTypeIPv6:
+	case nasMessage.PDUSessionTypeIPv4IPv6:
+		addrLen = 12 + 1
+	}
+	return addr, addrLen
+}
+
+func CreatePDUSession(ulNasTransport *nasMessage.ULNASTransport,
+	ue *UEContext,
+	fgc *Aio5gc,
+	pduSessionID int32,
+	smMessage []uint8,
+) (smContext *SmContext, err error) {
+	session := fgc.session
+
+	var (
+		snssai models.Snssai
+		dnn    string
+	)
+	// If the S-NSSAI IE is not included, select a default snssai
+	if ulNasTransport.SNSSAI != nil {
+		snssai = nasConvert.SnssaiToModels(ulNasTransport.SNSSAI)
+	} else {
+		snssai = ue.GetNssai()
+	}
+
+	dnnList := session.GetDnnList()
+	if ulNasTransport.DNN != nil {
+		if !slices.Contains(dnnList, ulNasTransport.DNN.GetDNN()) {
+			return nil, errors.New("[5GC] Unknown DNN requested")
+		}
+		dnn = ulNasTransport.DNN.GetDNN()
+
+	} else {
+		dnn = dnnList[0]
+	}
+
+	newSmContext := NewSmContext(pduSessionID)
+	newSmContext.SetSnssai(snssai)
+	dn, err := session.GetDataNetwork(dnn)
+	if err != nil {
+		return nil, err
+	}
+	newSmContext.SetDataNetwork(dn)
+
+	locationCopy := deepcopy.Copy(*ue.GetUserLocationInfo()).(models.NrLocation)
+	newSmContext.SetUserLocation(locationCopy)
+
+	n1smContent := ulNasTransport.PayloadContainer.GetPayloadContainerContents()
+	m := nas.NewMessage()
+	err = m.GsmMessageDecode(&n1smContent)
+	if err != nil {
+		return nil, errors.New("[5GC][NAS] GsmMessageDecode Error: " + err.Error())
+	}
+	if m.GsmHeader.GetMessageType() != nas.MsgTypePDUSessionEstablishmentRequest {
+		return nil, errors.New("[5GC][NAS] UL NAS Transport container message expected to be PDU Session Establishment Request but was not")
+	}
+	sessionRequest := m.PDUSessionEstablishmentRequest
+
+	newSmContext.SetPti(sessionRequest.GetPTI())
+	newSmContext.SetPduSessionType(sessionRequest.GetPDUSessionTypeValue())
+	newSmContext.SetSessionRule(session.GetSessionRules()[0])
+	newSmContext.SetDefQosQFI(uint8(1))
+	ip, err := tools.IncrementIP(session.lastAllocatedIP.String(), "10.0.0.0/8")
```

Comment:
this is not thread-safe, add a dedicated method to update lastAllocatedIP and returns the allocated IP
you should try running the aio5gc against packetrusher multi-ue